feat: integrating user and journal

d-e-v-esh opened the pull request:
## #61
This PR adds authorization on the backend using an auth middleware. Because of the auth middleware, a user must be logged in to `create`, `update`, or `delete` a journal. Further integration of users and journals was also added. Whenever a user creates a journal, that journal's `createdBy` field is populated by that user's Id.

### Changes

 - Added timestamps in journal schema. This handles `createdAt` and `updatedAt` fields automatically.
 - Added `logout` Mutation
 - Added authorization middleware
 - Added error handling in `createJournal` Mutation
 - Added error handling in `updateJournal` Mutation
 - Added boolean return type in `deleteJournal` Mutation
 - Integration of `user` and `journal` entities.
 
### Node Packages Added

[@graphql-tools/schema](https://www.npmjs.com/package/@graphql-tools/schema)
[graphql-middleware](https://www.npmjs.com/package/graphql-middleware)
[graphql-shield](https://www.npmjs.com/package/graphql-shield)

d-e-v-esh committed:
feat(journal-crud): timestamps
This commit removes the manual createdAt and updatedAt fields
and replaces it with timestamps which do the same job.
Now journals will have their createdAt and updatedAt populated
automatically.

d-e-v-esh committed:
feat(user-auth): add authorization middleware
This commit adds authorization middleware to the project
which is going to prevent people from creating, updating, and
deleting a journal without making an account.

d-e-v-esh committed:
feat(user-auth): add logout mutation
This commit adds the logout mutation to the backend.
Executing the logout function will destroy the session for that user
and delete the cookie saved in the user's browser.

d-e-v-esh committed:
refactor(journal-crud): simplify data entry
This commit removed the method of destructuring values from the
createJournal and updateJournal parameters and replaces it with
the rest operator.

d-e-v-esh committed:
feat(journal-crud): createJournal error handling
createJournal mutation will now throw an error when we try to
submit a journal with an issn that is already present in the database.

d-e-v-esh committed:
feat(journal-crud): updateJournal error handling
updateJounral will now throw an error when we try to submit a journal
with an ISSN that is already present in the database.

d-e-v-esh committed:
feat(journal-crud): add deleteJournal return type
This changes the return type of deleteJournal from String to Boolean.
If a journal is successfully deleted then it will return a true
otherwise it will return a false.

d-e-v-esh committed:
feat(journal-crud): updateJournal error handling
updateJournal will return an appropriate error
when we try to update a journal that is not present in the database.

d-e-v-esh committed:
refactor(journal-crud): change getCurrentUser query
In this commit, findOne() query is replaced byfindById()
which is used for the purpose of fetching the current user.

d-e-v-esh committed:
feat(journal-crud): integrate journal and user
Whenever a user creates a journal, that journal's createdBy field
is going to be populated by that user's ID and that journal's ID will be
inserted into that user's journal array.

CIPHERTron commented on src/models/Journal.js:
```diff
@@ -1,61 +1,58 @@
+    domainName: {
       type: String,
-      required: true,
+      required: false,
```
Shouldn't the `domainName` field be a **required** field?

CIPHERTron commented on src/models/Journal.js:
```diff
@@ -1,61 +1,58 @@
-    type: Date,
-    immutable: true,
-  },
+    policies: {
```
This looks good! 👍🏻

CIPHERTron commented on src/resolvers/journalResolver.js:
```diff
@@ -12,26 +13,63 @@ const journalResolver = {
+      } catch (error) {
+        if (error.code === 11000 && Object.keys(error.keyValue)[0] === "issn") {
+          return {
+            errors: [{ field: "issn", message: "issn already exists" }],
```
Change the error message to:
```
A journal with the same ISSN already exists
```

It would be even better if we use the string template literal & display the ISSN in the error message

CIPHERTron commented on src/resolvers/journalResolver.js:
```diff
@@ -12,26 +13,63 @@ const journalResolver = {
+      } catch (error) {
+        if (error.code === 11000 && Object.keys(error.keyValue)[0] === "issn") {
+          return {
+            errors: [{ field: "issn", message: "new issn already exists" }],
```
Same message as above!

CIPHERTron commented on src/index.js:
```diff
@@ -53,8 +57,10 @@ const startServer = async () => {
+    schema: applyMiddleware(
+      makeExecutableSchema({ typeDefs, resolvers }),
+      authMiddleware
+    ),
```
Loved this schema implementation using `makeExecutableSchema` 🔥

CIPHERTron reviewed (commented):
Just a few nitpicks @d-e-v-esh otherwise all good to go!

d-e-v-esh committed:
refactor(journal-crud): changed error message
In this commit, one of the error messages in
createJournal and updateJournal is updated.

d-e-v-esh commented:
All done 👍

CIPHERTron reviewed (approved):
LGTM
